### ISSUE TITLE
Can plugin be notified when netopeer2-cli subscribes for a certain notification

### ISSUE THREAD

**Issue by molisado** [sysrepo/sysrepo#2233]
I am working on a sysrepo plugin which implements the O-RAN M-Plane supervision mechanism, 
where the o-ran-supervision.yang defines the follow notification:  

```
module o-ran-supervision {
  notification supervision-notification {
    description
      "Notification to indicate that NETCONF management interface is up.";

  }
}
```

According to my understanding of the O-RAN M-Plane spec, the NETCONF server should start to periodically send the above notification to the NETCONF client once this client establishes the notification subscription. So my sysrepo plugin needs to be informed whenever a NETCONF client subscribes for supervision-notification.

I have 2 ideas for the solution:
1. Look for something like netconf-session-start notifcation (maybe netconf-subscription-start notification?)
After looking through ietf-netconf-notification.yang, nc-notifications.yang, notifications.yang, sysrepo-monitoring.yang,
the most similar one is /sysrepo-monitoring:sysrepo-state/module/subscriptions/notification-sub, as recommended by
https://github.com/sysrepo/sysrepo/issues/2198#issuecomment-719421749.
But this solution requires my plugin to keep polling the operational data instead of being notified.

2. My plugin subscribes for the RPC create-subscription
I try this solution before seeing the post (https://github.com/CESNET/netopeer2/issues/375#issuecomment-450815108), which states that the create-subscription RPC cannot be handled by a plugin callback. However, the callback of my plugin for RPC create-subscription did get invoked when I test with netopeer2-cli using "subscribe --filter-xpath /o-ran-supervision:supervision-notification" command. But there is a problem, the callback in my plugin cannot distinguish which notification subscription it is. I think this information is provided by the "filter" input of the RPC, but the "input_cnt" argument of the callback is 0?

The code of my plugin (just for test):
```
rpc_cb(sr_session_ctx_t *session, const char *path, const sr_val_t *input, const size_t input_cnt, sr_event_t event, uint32_t request_id, sr_val_t **output, size_t *ouput_cnt, void *private_data)
{
    printf("input_cnt(%zu)\n", input_cnt);
    ...
}

sr_plugin_init_cb(sr_session_ctx_t *sess, void **private_ctx)
{
    sr_rpc_subscribe(sess, "/notifications:create-subscription", rpc_cb, NULL, 0, SR_SUBSCR_CTX_REUSE, &subscription);
    ...
}
```

The version of each component is:
libyang: 1.0.184
libnetconf2: 1.1.26
sysrepo: 1.4.66
Netopeer2: 1.1.34

**Comment by michalvasko** [comment]
1. You are right about the solution and its problems, nothing for me to add.
2. That should be possible, what I wrote was true for the legacy *sysrepo*/*netopeer2*, the current one allows this. If the `filter` is not present, it means all the notifications were subscribed to. Also, you must check `stream` and your notification was also subscribed to if it is `NETCONF` (all the streams/notifications) or `o-ran-supervision` (all notifications from this module).

**Comment by molisado** [comment]
> 2. That should be possible, what I wrote was true for the legacy _sysrepo_/_netopeer2_, the current one allows this. If the `filter` is not present, it means all the notifications were subscribed to. Also, you must check `stream` and your notification was also subscribed to if it is `NETCONF` (all the streams/notifications) or `o-ran-supervision` (all notifications from this module).

Thanks for your reply. But I am still confused about the behavior of my sysrepo plugin callback for create-subscription RPC.
Below is my plugin functions, which is mentioned before, but here with a little more detail to display the RPC parameters:
```
static int
rpc_cb_create_subscription(sr_session_ctx_t *session, const char *path, const sr_val_t *input, const size_t input_cnt,
        sr_event_t event, uint32_t request_id, sr_val_t **output, size_t *output_cnt, void *private_data)
{
    size_t i;
    
    printf("enter: input_cnt(%lu)\n", input_cnt);
    
    for (i = 0; i < input_cnt; i++) {
        printf("input[%lu]: type(%d), xpath(%s).\n", i, input[i].type, input[i].xpath);
        if (17 == input[i].type) {
            printf("input[%lu]: data(%s).\n", i, input[i].data.string_val);
        }
    }
    
    *output_cnt = 0;
    
    return SR_ERR_OK;
}
```
To test my plugin, I use netopeer2-cli to perform the following 3 subscribe commands:
```

> subscribe --filter-xpath /ietf-hardware:hardware-state-change
OK
> 
> 
> 
> 
> subscribe --filter-xpath /ietf-hardware:hardware-state-oper-enabled
OK
> 
> 
> 
> subscribe --filter-xpath /ietf-hardware:hardware-state-oper-disabled
nc ERROR: Session 1: SSH channel poll error (Socket error: disconnected).
cli_send_recv: Failed to receive a reply.
>
```

And the messages displayed by my plugin are:
```
enter: input_cnt(0)

enter: input_cnt(0)

enter: input_cnt(2)
input[0]: type(17), xpath(/notifications:create-subscription/stream).
input[0]: data(NETCONF).
input[1]: type(22), xpath(/notifications:create-subscription/filter).
```

Moreover, the 3rd subscribe command causes netopeer2-server to crash due to segmentation fault, where the messages are:
```

[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: SR: Published event "rpc" "/notifications:create-subscription" with ID 1 priority 0 for 2 subscribers.
[INF]: SR: Processing "/notifications:create-subscription" "rpc" event with ID 1 priority 0 (remaining 2 subscribers).
[INF]: SR: Successful processing of "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: SR: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: NP: Session 1: thread 1 event new RPC.

[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: SR: Published event "rpc" "/notifications:create-subscription" with ID 2 priority 0 for 2 subscribers.
[INF]: SR: Processing "/notifications:create-subscription" "rpc" event with ID 2 priority 0 (remaining 2 subscribers).
[INF]: SR: Successful processing of "rpc" event with ID 2 priority 0 (remaining 1 subscribers).
[INF]: SR: Event "rpc" with ID 2 priority 0 succeeded.
[INF]: NP: Session 1: thread 0 event new RPC.

[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: SR: Published event "rpc" "/notifications:create-subscription" with ID 3 priority 0 for 2 subscribers.
[INF]: SR: Processing "/notifications:create-subscription" "rpc" event with ID 3 priority 0 (remaining 1 subscribers).
Segmentation fault (core dumped)
```

In summary, I think there are two problems here:

1. The RPC input parameters provided to the plugin callback is "non-deterministic"?
The 3 subscribe commands I tested should all provide "stream" and "filter" input parameters to my plugin callback, but the test results are not like that.

2. Why netopeer2-server crashed after my plugin callback returned?

**Comment by michalvasko** [comment]
Right, I believe you are affected by a bug that was fixed by 134ae643d3afa13867c1fd748a72514edc3a6c4b. Please use the current `devel` that should have this fixed.

**Comment by molisado** [comment]
> Right, I believe you are affected by a bug that was fixed by [134ae64](https://github.com/sysrepo/sysrepo/commit/134ae643d3afa13867c1fd748a72514edc3a6c4b). Please use the current `devel` that should have this fixed.

Thanks for your reply. Changing to the `devel` branch did improve the situation.
Now the "create-subscription" RPC callback seems to provide the RPC input.
But the next problem is that I cannot retrieve the value of the "filter" RPC input.

Please refer to the my callback to "create-subscription" RPC below, which simply display each input data according to its type:
```
static int
rpc_cb_create_subscription(sr_session_ctx_t *session, const char *path, const sr_val_t *input, const size_t input_cnt,
        sr_event_t event, uint32_t request_id, sr_val_t **output, size_t *output_cnt, void *private_data)
{
    size_t i;
    
    printf("enter: input_cnt(%lu)\n", input_cnt);
    
    for (i = 0; i < input_cnt; i++) {
        printf("input[%lu]: type(%d), xpath(%s).\n", i, input[i].type, input[i].xpath);
        if (SR_STRING_T == input[i].type) {
            printf("input[%lu]: data.string_val(%s).\n", i, input[i].data.string_val);
        } else if (SR_ANYXML_T == input[i].type) {
            printf("input[%lu]: data.string_val(%s)/anyxml_val(%s)/anydata_val(%s).\n",
                      i,
                      input[i].data.string_val ? input[i].data.string_val : "NULL",
                      input[i].data.anyxml_val ? input[i].data.anyxml_val : "NULL",
                      input[i].data.anydata_val ? input[i].data.anydata_val : "NULL");
        }
    }
    
    *output_cnt = 0;
    
    return SR_ERR_OK;
}
```

The above callback is tested with 2 kinds of netopeer2-cli subscribe commands:

1. with --filter-xpath parameter:
```
> subscribe --filter-xpath /o-ran-supervision:supervision-notification
OK
>
```

The messages displayed by my callback:
```
enter: input_cnt(2)
input[0]: type(17), xpath(/notifications:create-subscription/stream).
input[0]: data.string_val(NETCONF).
input[1]: type(22), xpath(/notifications:create-subscription/filter).
input[1]: data.string_val()/anyxml()/anydata().
```

2. with --stream parameter:
```
> subscribe --stream o-ran-supervision
OK
>
```

The messages displayed by my callback:
```
enter: input_cnt(1)
input[0]: type(17), xpath(/notifications:create-subscription/stream).
input[0]: data.string_val(o-ran-supervision).
```

Although the type for the `filter` input is SR_ANYXML_T, but the content of input.data.anyxml_val is empty.
In case I might miss anything, the content of input.data.anydata_val & string_val are also printed, which are both empty, too.
So where can I retrieve the value of the `--filter-xpath` input?
Maybe my perception that the `filter` input to the "create-subscription" RPC is equivalent to the content of `--filter-xpath` is wrong?

**Comment by michalvasko** [comment]
Right, XPath filter is encoded [as an attribute](https://tools.ietf.org/html/rfc6241#section-8.9.5.1), which is not really supported by the `sr_val_t` API. But if you use `sr_rpc_subscribe_tree()` instead, you will get the access to the exact received RPC, including its attributes (metadata).

**Comment by molisado** [comment]
Thank you so much.
sr_rpc_subscribe_tree() did solve my problem.

**Comment by michalvasko** [comment]
> The callback sr_rpc_tree_cb() gets from sysrepo just data we used for sr_rpc_subscribe_tree()

That is not true, `input` is a data tree, not just a single node. Just call `lyd_print_file()` on it and you will see all that all the information is there.
